draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -356,869 +524,832 @@ document are to be interpreted as described in {{RFC2119}}.
-`q=2 mod 3`.
-
-**Examples**
+## Encodings for Montgomery curves
 
-- `y^2 = x^3 + 1`
+### Elligator2 Method {#elligator2}
 
-**Algorithm**: map2curve_bf
+The map2curve_elligator2(alpha) implements the Elligator2 {{Elligator2}} for
+curves defined by y^2 = x^3 + A\*x^2 + B\*x such that A\*B\*(A^2 - 4\*B) != 0.
+In particular, this method applies to the Montgomery curves y^2 = x^3 + A\*x^2 + x
+setting B=1.
 
-Input:
+Preconditions: A Montgomery curve  such that A!=0.
 
- - `alpha`: an octet string to be hashed.
- - `B`: the constant from the Weierstrass curve.
+Input: alpha, an octet string to be hashed.
 
-Output:
+Constants: A and B, the parameters of the curve; N, a non-square in F.
 
- - `(x, y)`: a point in E.
+Output: (x, y), a point on E.
 
 Operations:
 
 ~~~
-1. u = hash2base(alpha)
-2. x = (u^2 - B)^((2 * q - 1) / 3)
-3. Output (x, u)
+1.   u = hash2base(alpha)
+2.  x1 = -A / (1 + N * u^2)
+3. gx1 = x1^3 + A * x1^2 + B * x1
+4.  x2 = -x1 - A
+5. gx2 = x2^3 + A * x2^2 + B * x2
+6.   e = gx1^((q - 1) / 2)
+7. If is_square(gx1), set x = x1 and y = -e * sqrt(gx1)
+8. If is_square(gx2), set x = x2 and y = -e * sqrt(gx2)
+9. Output (x, y)
 ~~~
 
-**Implementation**
+#### Implementation
 
-The following procedure implements the Boneh-Franklin's algorithm in a
-straight-line fashion.
+The following procedure implements elligator2 algorithm in a straight-line
+fashion.
 
 ~~~
-map2curve_bf(alpha)
-
-Input:
-
- alpha: an octet string to be hashed.
- B    : the constant from the Weierstrass curve.
-
-Output:
-
- (x, y): a point in E
-
-Precomputations:
+map2curve_elligator2(alpha)
+Input: alpha, an octet string to be hashed.
+Output: (x, y), a point on E.
 
-1.  c = (2 * q - 1) / 3    // Integer arithmetic
+Constants:  
+1. c1 is an non-square in F.
+2. c2 = (q - 1) / 2      // Integer arithmetic
 
 Steps:
+1.    u = hash2base(alpha)
+2.   x1 = u^2
+3.   x1 = c1 * x1        
+4.   x1 = x1 + 1
+5.   x1 = 1 / x1
+6.   x1 = A * x1
+7.   x1 = -x1            // x1 = -A / (1 + N * u^2)
+8.  gx1 = x1 + A
+9.  gx1 = gx1 * x1
+10. gx1 = gx1 + B
+11. gx1 = gx1 * x1       // gx1 = x1^3 + A * x1^2 + B * x1
+12.  x2 = -x1 - A        // x2 = -x1 - A
+13. gx2 = x2 + A
+14. gx2 = gx2 * x2
+15. gx2 = gx2 + B
+16. gx2 = gx2 * x2       // gx2 = x2^3 + A * x2^2 + B * x2
+17.   e = is_square(gx1, q)
+18.   x = CMOV(x2, x1, e)    // If e=True, x=x1, else x=x2
+19.  gx = CMOV(gx2, gx1, e)  // If e=True, x=gx1, else x=gx2
+20.  y2 = sqrt(gx, q)
+22.  y1 = -y2
+22.   y = CMOV(y2, y1, e)    // If e=True, y=y1, else y=y2
+23. Output (x, y)
+~~~
+
+## Encodings for twisted Edwards curves
+
+### Elligator2 Method {#ell2edwards}
+
+The map2curve_ell2edwards(alpha) implements an adaptation of Elligator2
+{{Elligator2}} for twisted Edwards curves defined by A\*x^2 + y^2 = 1 + D\*x^2y^2
+over a field F.
+
+Preconditions: A Twisted Edwards curve.
+
+Input: alpha, an octet string to be hashed.
+
+Constants: A and D, the parameters of the curve; c0, a non-square in F;
```

Comment:
nit: the style of this constants list does not match that of previous sections

Reply:
Changed.